### ISSUE TITLE
Is it possible to add LoRA on specific head?

### ISSUE THREAD

**Issue by SpeeeedLee** [huggingface/peft#2293]
### Feature request

Could I add LoRA only to some selected heads on the model?
I read some documentation [here](https://huggingface.co/docs/peft/developer_guides/custom_models), but am still not sure about how to implement my goal.

### Motivation

Current LoRA Config can allow users to decide where matrices to add LoRA, a more fine-grained control on which heads to add LoRA would be beneficial for the developers.

### Your contribution

I would appreciate some tips on how to implement this.

**Comment by d-kleine** [comment]
With `target_modules`, it's described here: https://huggingface.co/docs/peft/developer_guides/custom_models#multilayer-perceptron

```py
config = LoraConfig(
    target_modules=["seq.0", "seq.2"], # use the layer names according to the model you are using
    modules_to_save=["seq.4"],
)
```

You can retrieve the name and type of each layer of your model with this code:

```py
for n, m in base_model.named_modules(): # replace `base_model` with the variable your pretrained model is stored in
    print((n, type(m)))
```

**Comment by SpeeeedLee** [comment]
Hi, @d-kleine, thanks for the reply. I am thinking more about something like adding LoRA on some attention heads only, which means maybe my target might be:
First `head-dim` columns of ``layers.0.self_attn.q_proj``, i.e, the first attention head of Query matrix. Besides adding LoRA, I am also curious about how can I freeze some heads and fine-tune some selected ones only.

**Comment by BenjaminBossan** [comment]
It is not possible to target specific heads. The issue is that the weights of all heads are combined into a single `nn.Linear` weight, so if we apply LoRA to it, it will affect all the weights.

**Comment by SpeeeedLee** [comment]
Thanks for the prompt reply, @BenjaminBossan. 
Yes, I understand this issue and am wondering whether it possible for me to write a custom code to separate each head to a single `nn.Linear` weight. If so, then I can selectively fine-tune some heads or adding LoRA on them.

I found some possible approaches like [this previous issue](https://github.com/huggingface/transformers/issues/33928), where SAM's Q, K, and V are successfully separated. This might be used in my case, where I want to separate each head out?

**Comment by BenjaminBossan** [comment]
> wondering whether it possible for me to write a custom code to separate each head to a single `nn.Linear` weight. If so, then I can selectively fine-tune some heads or adding LoRA on them.

That is possible, it means that you have to implement the whole transformer attention module for yourself and you might be missing out on some optimizations (flash attention, caching).

Alternatively, you might be able to write a custom LoRA layer that, say, masks out the heads that should not be touched, and [register it with the PEFT dispatcher](https://huggingface.co/docs/peft/v0.14.0/en/developer_guides/custom_models#experimental-support-for-dynamic-dispatch-of-custom-modules-in-lora) to be applied to the whole attention module, e.g. [`LlamaAttention`](https://github.com/huggingface/transformers/blob/24c91f095fec4d90fa6901ef17146b4f4c21d0a3/src/transformers/models/llama/modeling_llama.py) if that's what your model is using.

**Comment by SpeeeedLee** [comment]
Thanks, @BenjaminBossan.

> it means that you have to implement the whole transformer attention module for yourself 

I tried to do this on `LlamaAttention`. However, before any fine-tuning, the resulting Llama perform slightly bad than the original one. It must be something wrong with my implementation. (I am using the latest transformers)
Would it be possible to provide some hints/example code on this?
Thanks a lot.

**Comment by BenjaminBossan** [comment]
You mean that using the same weights with your implementation, the performance already is degraded? Yes, this most likely means there is a bug somewhere. You could paste your implementation here and mark the parts of the code that you changed, and I can take a look.

**Comment by SpeeeedLee** [comment]
Hi, @BenjaminBossan, I've been able to separate out each head-related submatrices from the original concatenated Q, K, V, O matrices now!
The separated LM has completely the same behavior as the original one.
Thank you a lot for your help on this.

However, I still have a small issue: the separated LM will not have the correct pretrained weights after initialization, and I will need to write an additional code to load the correct weights manually.
Do you have any suggestions on this?

Thanks in advance.

**Comment by BenjaminBossan** [comment]
Nicely done getting the implementation to work correctly. As for loading the weights, it's hard to say in the abstract without seeing your code. In general, I'd try:

1. Load the correct weights before separating the heads, so that they start with the right weights.
2. Iterate through the state dict, filter the matrices, and split them according to the same logic you already have. Ensure that the key names are adjusted to fit your new module.

**Comment by RayneYael** [comment]
Hi @SpeeeedLee,

I found your work on fine-tuning specific heads very interesting! I'm currently attempting to do something similar with Qwen2.5-VL-3B, but I'm running into OOM issues.

In my initial attempt, I used backward hooks to apply a mask (setting gradients of non-candidate heads to 0), but this approach seems memory-intensive. I’m very curious about your implementation strategy for selectively fine-tuning heads efficiently.

Any insights or code snippets would be greatly appreciated!
